test/RewardsStreamerMP.t.sol

```diff
@@ -1102,6 +1102,89 @@ contract UnstakeTest is StakeTest {
         );
     }
 
+    function test_UnstakeOneAccountAndAccruedMP() public {
+        test_StakeOneAccount();
+
+        // wait for 1 year
+        uint256 currentTime = vm.getBlockTimestamp();
+        vm.warp(currentTime + (365 days));
+
+        streamer.updateGlobalState();
+        streamer.updateUserMP(alice);
+
+        checkStreamer(
+            CheckStreamerParams({
+                totalStaked: 10e18,
+                totalMP: 20e18, // total MP must have been doubled
+                totalMaxMP: 40e18,
+                stakingBalance: 10e18,
+                rewardBalance: 0,
+                rewardIndex: 0,
+                accountedRewards: 0
+            })
+        );
+
+        // unstake half of the tokens
+        _unstake(alice, 5e18);
+
+        checkStreamer(
+            CheckStreamerParams({
+                totalStaked: 5e18, // 10 - 5
+                totalMP: 10e18, // 20 - 10 (5 initial + 5 accrued)
+                totalMaxMP: 20e18,
+                stakingBalance: 5e18,
+                rewardBalance: 0,
+                rewardIndex: 0,
+                accountedRewards: 0
+            })
+        );
+    }
+
+    function test_UnstakeOneAccountWithLockUpAndAccruedMP() public {
+        test_StakeOneAccountWithMinLockUp();
+
+        uint256 stakeAmount = 10e18;
+        uint256 lockUpPeriod = streamer.MIN_LOCKING_PERIOD();
+        // 10e18 is what's used in `test_StakeOneAccountWithMinLockUp`
+        uint256 expectedBonusMP = _calculateBonusMP(stakeAmount, lockUpPeriod);
+
+        // wait for 1 year
+        uint256 currentTime = vm.getBlockTimestamp();
+        vm.warp(currentTime + (365 days));
+
+        streamer.updateGlobalState();
+        streamer.updateUserMP(alice);
+
+        checkStreamer(
+            CheckStreamerParams({
+                totalStaked: stakeAmount,
+                totalMP: (stakeAmount + expectedBonusMP) + stakeAmount, // we do `+ stakeAmount` we've accrued
+                    // `stakeAmount` after 1 year
+                totalMaxMP: stakeAmount * streamer.MAX_MULTIPLIER() + expectedBonusMP,
+                stakingBalance: 10e18,
+                rewardBalance: 0,
+                rewardIndex: 0,
+                accountedRewards: 0
+            })
+        );
+
+        // unstake half of the tokens
+        _unstake(alice, 5e18);
+        expectedBonusMP = _calculateBonusMP(5e18, lockUpPeriod);
+
+        checkStreamer(
+            CheckStreamerParams({
+                totalStaked: 5e18,
+                totalMP: (5e18 + expectedBonusMP) + 5e18,
+                totalMaxMP: 5e18 * streamer.MAX_MULTIPLIER() + expectedBonusMP,
+                stakingBalance: 5e18,
+                rewardBalance: 0,
+                rewardIndex: 0,
+                accountedRewards: 0
+            })
+        );
```

Comment:
This passes, but I was wondering something.

In this scenario, I've staked for a year, and I locked up my funds for 3 months.
For locking up my funds, I've received a higher initial MP and my max MP has increase as well.
Cool.

Now a year has passed, I'm unstaking half of my funds, which means I should also loose half of my **initial** MP.
But should I also loose half of the bonus MP for which I have locked up my funds for 3 months initially?

I guess what I'm unsure about is, should I really loose the bonus MP for which I've done my work?
I might no longer stake the 10e18 tokens, but I did for 3 months, for which I'm entitled to that bonus.
Right now that is taken away from me when I partially unstake.

I wonder if we should keep, at least as long as the user is staked in the system.

Or maybe this is total nonsense.

Curious about your thoughts @gravityblast @3esmit

Reply:
Okay, actually on a second thought I don't think this is a good idea.

If we'd keep MPs for the lock up that was done, we'd have to keep track of all lock ups per stake to make sure this checks out.